src/scheduler/constant.h

```diff
@@ -168,6 +168,18 @@ enum resval_cmpflag
 	CMP_CASE,
 	CMP_CASELESS
 };
+
+enum thread_task_type
```

Comment:
To be honest, I don't think any other PBS daemon will benefit from multi-threading like this. PBS Server doesn't do heavyweight calculations like the scheduler does, most of its cpu time is spent sending and receiving requests. The mom daemon just launches jobs and monitors them, not much to process there either. pbs_comm has its own thing already.

So, given how using callbacks degrades performance, we should only use them if they'll benefit us in a significant way, and I just don't see that ROI. Please let me know what you think.

Reply:
You have a fine implementation of thread pool in place. It would be a shame if other components can not make use of it without making significant changes to it.
I understand server may not have as many places it can use a thread pool as scheduler does, but, there could be cases where it might need it in future (like sending mom hooks, clearing up job-history, maybe implement a notification service or do other things).
One of the reasons deleting jobs is slow is because mom is trying to scan through the proc file system for every session ID it knows (serially), with thread-pool it can use threads to do that. Mom could also use this to monitor jobs as well.

IIRC, you mentioned that scheduler takes around half the time to process a million jobs when this thread implementation is being used (roughly from 6mins to 3 mins), but when you use function pointers it drops by 40 seconds. While 40 seconds is a lot of degradation for a 3 min cycle, but I think we have still improved it by more than 2 mins out of 6 mins which is significant.

Reply:
hmm .. so we don't really need to change the algorithm to use callbacks in order to make it generic, right? we can just add more tasks to the task list and call their corresponding task handlers, which could be non-scheduler. So, if we ever have a need for it, we can just put the code in a common library instead of the scheduler. Does that work?

Reply:
I don't think just adding the tasks to the list and calling their respective methods will do it. a non-scheduler component will then be compiling with references present in sched library.

Reply:
ok ya, you are correct .. I don't know man, I really don't want to make a change which will hurt performance when the whole purpose of this PR is to improve performance. If some other part of PBSPro needs multi-threading, it'll be better for performance to just create the infrastructure for that part separately, it's really not that much code, most of the code change in this PR is the task handlers themselves. We should also keep in mind that, depending on the use case, the other parts of PBS might need a different multi-threading approach than the one used here with the worker/result queues. So, I'd like to keep this as it as for now. If you still think we should go the callback route, then we can ask others for opinions.

Reply:
Ok, do you have the function pointer implementation of yours that you can share. It will be interesting to see why is function pointer slower than managing a switch statement.
I agree that since our discussion isn't moving anywhere, It would be better to seek opinions from others.

Reply:
I agree with Arun here, in general (without actually seeing the specific implementation or considerations here). It would be great if you can make the thread pool code generic (and into a library) that can then be utilized by anybody (daemons or not) wanting to use a thread pool. I don't know what that would be the impact on the timeline for this PR, but we should consider it. If we cannot make that refactoring within this PR's timeframe, we could file another refactoring RFE to do it shortly after?